vercel/edge-runtime
# feat(feature-detector): introduces edge signature detection

### 📖 What's in there?

Here is the first stone of the feature detector package: detecting exported function with an edge signature.
This PR includes:
- the new `feature-detector` package
- the `hasEdgeSignature(filePath: string): boolean` function, fully tested. 
- documentation page (giant WIP with a useful example).
- README, redirecting to the documentation page.

### 🧪 How to test?

The package is fully covered with unit tests: `pnpm --filter feature-detector test -- --coverage`

You can also try it out on a real file:
1. `pnpm --filter feature-detector build`
2. Create a file named `test.js` (or .ts)
3. start `node` REPL, and try it:
   ```shell
   > const { hasEdgeSignature } = require('./packages/feature-detector/dist')
   undefined
   > hasEdgeSignature('./test.js')
   false
4. change the file content and check its compliance with `hasEdgeSignature()` again.


Documentation can be seen at http://localhost:3000/packages/feature-detector after starting the doc site with `pnpm --filter docs dev`
You can also review it [here](https://edge-runtime-git-damien-ec-544-build-a-minimal-ts-morph-61fc56.vercel.sh/packages/feature-detector)

### <picture data-single-emoji=":this:" title=":this:"><img class="emoji" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..kRl7soU3nA940o4U.2oFyp4zfoNI1Hm0-F_OcIyY66WYrpMbvLTmnK_3oWjEFWhOTaNYfSBHn4Yx4dirRaAf4mmOwPvtQUeAis4nr6pYy7oXt3xAiMzhXm2NEAfmjdUcebRA.BvFwmzy6GxjiRPIj1WODsQ" alt=":this:" width="20" height="auto" align="absmiddle"></picture> Notes to reviewers

Because `Response.json()` static method is not supported in Typescript default types ([lib.dom.d.ts](https://github.com/microsoft/TypeScript/blob/4b52d3a61a05a7c2b3d03a8e1c801560232f3194/lib/lib.dom.d.ts#L11782-L11787)), I had to overload default types with an augmented definition.

This implies redefining `Response`, but also discarding all other default types with `/// <reference no-default-lib="true"/>`.
As a consequence, we're loosing `Promise`, `fetch()`, and all other build-in types...

Hence why I had to download the entire es2022 + DOM type definitions, crunch them in a file (with no comments to make it lighter), and load it into the temporary ts-morph project used to analyze user code.
